cxx-qt-lib/src/qurl.rs

```diff
@@ -0,0 +1,242 @@
+        };
+    };
+}
+
+impl QUrl {
+    /// `QUrl` is not constructible via `new`.
+    /// Instead, use the [`let_qurl!`] macro.
+    pub fn new<T: Private>() -> Self {
+        unreachable!()
+    }
+
+    /// Create a new Rust Url from this QUrl.
+    /// This is a copy operation so any changes will not propagate to the original QUrl.
+    pub fn to_rust(&self) -> Url {
+        let mut data = String::new();
+        // Safety:
+        //
+        // Given that a QUrl can only be constructed using [`let_qurl!`] macro,
+        // it is safe to assume that self is a valid QUrl reference which makes this
+        // function call safe.
+        unsafe { qurl_to_rust_string(self, &mut data) };
+
+        Url::from_string(data)
+    }
+}
+
+#[doc(hidden)]
+#[repr(C)]
+pub struct StackQUrl {
+    // Static assertions in cxx_qt.cpp validate that this
+    // is large enough and aligned enough.
+    space: MaybeUninit<[usize; 1]>,
+}
+
+#[allow(missing_docs)]
+impl StackQUrl {
+    pub fn new() -> Self {
+        StackQUrl {
+            space: MaybeUninit::uninit(),
+        }
+    }
+
+    /// # Safety
+    ///
+    /// Calling this function twice on the same StackQUrl is unsafe
+    /// and leads to undefined behaviour. It is therefore recommended
+    /// to not use this function directly and instead use the [`let_qurl!`]
+    /// macro which ensures that safe behaviour.
+    pub unsafe fn init(&mut self, value: &Url) -> Pin<&mut QUrl> {
+        let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<QUrl>>();
+        let value = &value.deref().data;
+        qurl_init_from_string(this, value.as_ptr(), value.len());
+        Pin::new_unchecked(&mut *this.as_mut_ptr())
+    }
+}
+
+impl Default for StackQUrl {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Drop for StackQUrl {
+    fn drop(&mut self) {
+        // # Safety
+        //
+        // This simply calls ~QUrl on self.space which is safe as long
+        // as self.space contains a valid QUrl. Using the [`let_qurl!`]
+        // macro guarantees that this will be the case.
+        unsafe {
+            let this = &mut *self.space.as_mut_ptr().cast::<MaybeUninit<QUrl>>();
+            qurl_drop(this);
+        }
+    }
+}
+
+// Safety:
+//
+// The code in this file ensures that QUrl can only ever be allocated
+// on the stack in pinned form which avoids the pitfalls of trying to
+// move this type that has a non-trivial move constructor.
+unsafe impl ExternType for QUrl {
+    type Id = type_id!("QUrl");
+    type Kind = cxx::kind::Opaque;
+}
+
+pub struct UrlImpl {
+    pub data: String,
+    // fragment: Option<String>,
+    // host: Option<String>,
+    // password: Option<String>,
+    // path: Option<String>,
+    // port: i32,
+    // query: Option<String>,
+    // scheme: Option<String>,
+    // userName: Option<String>,
+}
+
+#[repr(C)]
+pub struct Url {
+    pub d: Box<UrlImpl>,
+}
+
+impl Url {
```

Comment:
Should we have a `as_string() -> &String` method here? So they UrlImpl doesn't need to be used in Rust?